<a href="https://colab.research.google.com/github/ricvazquez/ColabFiles/blob/main/Actividad_Calificada_07_Datos_Duplicados_273509.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | |
| **Matrícula** | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [ ]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [ ]:
# Escribe tu código aquí
df_h = pd.DataFrame(hoteles)
dubs = df_h.copy()
# Sacamos los duplicados con la regla en first para los primeros
dubs['keep_first'] = dubs.duplicated(keep='first')
# Sacamos los duplicados con la regla last para los ultimos
dubs['keep_last'] = dubs.duplicated(keep='last')
# Sacamos los duplicados con la regla en false
dubs['keep_false'] = dubs.duplicated(keep=False)
print(dubs['keep_first'].value_counts())
print(dubs['keep_last'].value_counts())
print(dubs['keep_false'].value_counts())

keep_first
False    87396
True     31994
Name: count, dtype: int64
keep_last
False    95567
True     23823
Name: count, dtype: int64
keep_false
False    96536
True     22854
Name: count, dtype: int64


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [ ]:
# Escribe tu código aquí
# Primero sacamos el promedio cuando dejamos primeros duplicados
kf = dubs['keep_first'].sum()/dubs['keep_first'].count() * 100
print('Porcentaje de Keep First', round(kf,2),'%')

#Sacamos el promedio de cuando dejamos los ultimos duplicados
kl = dubs['keep_last'].sum()/dubs['keep_last'].count() * 100
print('Porcentaje de Keep Last', round(kl,2),'%')

#Sacamos el prmedio de los duplicados en general
kf = dubs['keep_false'].sum()/dubs['keep_false'].count() * 100
print('Porcentaje de Keep False', round(kf,2),'%')
print('Grupos repetidos', dubs['keep_false'].sum())

Porcentaje de Keep First 26.8 %
Porcentaje de Keep Last 19.95 %
Porcentaje de Keep False 19.14 %
Grupos repetidos 22854


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [ ]:
# Escribe tu código aquí
copy = hoteles.copy()
# Sacamos los duplicados en first y creamos una columna es duplicado
copy['es_duplicado'] = copy.duplicated(keep='first')
# Agrupamos por hotel y sacamos la sumatoria de cada hotel y lo guardamos en una variable
hotels = copy.groupby('hotel')['es_duplicado'].sum()
# print(hotels)
# Recorremos hotels
for i,v in hotels.items():
  print(i, round(v/copy['es_duplicado'].count()*100,2),'%')

City Hotel 21.7 %
Resort Hotel 5.1 %


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

Por lo visto es un problema grave solo en el city hotel ya que tiene 1/5 parte de datos duplicados, en cambio resort hotel solo tiene 5 %

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [ ]:
# Escribe tu código aquí

# Creamos la copia nueva
copy = hoteles.copy()
# Guardamos los lo del duplicated en una columna
copy['es_duplicado'] = copy.duplicated(keep=False)
# Nos quedamoscon solo las true
copy = copy[copy['es_duplicado'] == True]
#Acomodamos como lo pide arriba
copy = copy.sort_values(by=['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adults', 'country',
                            'market_segment', 'deposit_type', 'adr', 'reservation_status'])
copy.head(12)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,es_duplicado
40643,City Hotel,1,0,2015,August,32,4,0,1,2,...,NaN,NaN,0,Transient,0.0,0,0,Canceled,2015-08-04,True
40644,City Hotel,1,0,2015,August,32,4,0,1,2,...,NaN,NaN,0,Transient,0.0,0,0,Canceled,2015-08-04,True
40772,City Hotel,0,0,2015,August,32,7,0,2,2,...,14.0,NaN,0,Transient,75.0,0,1,Check-Out,2015-08-09,True
40802,City Hotel,0,0,2015,August,32,7,0,2,2,...,14.0,NaN,0,Transient,75.0,0,1,Check-Out,2015-08-09,True
40821,City Hotel,0,0,2015,August,32,8,0,1,2,...,9.0,NaN,0,Transient,89.0,0,1,Check-Out,2015-08-09,True
40838,City Hotel,0,0,2015,August,32,8,0,1,2,...,9.0,NaN,0,Transient,89.0,0,1,Check-Out,2015-08-09,True
76792,City Hotel,0,0,2015,August,33,10,1,0,2,...,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-08-11,True
76793,City Hotel,0,0,2015,August,33,10,1,0,2,...,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-08-11,True
76794,City Hotel,0,0,2015,August,33,10,1,0,2,...,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-08-11,True
41067,City Hotel,0,0,2015,August,33,11,0,1,1,...,NaN,38.0,0,Transient-Party,88.0,0,0,Check-Out,2015-08-12,True


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [ ]:
# Escribe tu código aquí
# Sacamos las repetidas y cuantas veces sin nulos
tam = df_h.value_counts(dropna=False)
# Dejamos solo los grupos mayores a dos
grupos = tam[tam >= 2]
# Los acomodamos de menor a mayor
grupos = grupos.sort_values(ascending=False)
# Agarramos los primeros 10 y renombramos las columnas para que se entienda
# 2 repetidos aparece 5034 veces
# 3 repetidos aparece 969 etc
tamg = grupos.value_counts().head(10).rename_axis('tamaño').reset_index(name='cantidad')
print(tamg)
print('El mas grande', grupos.max())

   tamaño  cantidad
0       2      5034
1       3       969
2       4       497
3       5       275
4       6       171
5      10       123
6       7       115
7       8       103
8      20        97
9       9        93
El mas grande 180


### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [ ]:
# Escribe tu código aquí
df_h2 = hoteles.copy()
# Creamos la columna
df_h2['es_duplicado'] = df_h2.duplicated(keep=False)
# Dejamos solo los true
df_h2 = df_h2[df_h2['es_duplicado'] == True]
# Imprimimos las proporciones
print(df_h2['market_segment'].value_counts(normalize = True))
print(df_h2['deposit_type'].value_counts(normalize = True))
print(df_h2['is_canceled'].value_counts(normalize = True))


market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64
deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64
is_canceled
1    0.583867
0    0.416133
Name: proportion, dtype: float64


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

Hay muchas filas que parecen estar duplicadas al ser totalmente iguales cada uno de sus registros, posiblemente una doble captura por error

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

Al tener NaN ambas, el duplicated lo sigue tomando como duplicado y lo toma en cuenta, a diferencia de cuando usamos calue counts(dropna = True) el cual los maneja diferente

**Respuesta 2.4:**

1.- Hay muchas filas que parecen estar duplicadas al ser totalmente iguales cada uno de sus registros, posiblemente una doble captura por error

2.- Al tener NaN ambas, el duplicated lo sigue tomando como duplicado y lo toma en cuenta, a diferencia de cuando usamos calue counts(dropna = True) el cual los maneja diferente

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [ ]:
# Escribe tu código aquí
hoteles_limpio = hoteles.copy()
# Borramos duplicados
hoteles_limpio = hoteles_limpio.drop_duplicates(keep=False)
# Limpiamos index
hoteles_limpio = hoteles_limpio.reset_index(drop=True)
# Imprimimos cuantos tenia antes, cuantos despues y la diferencia
print('Filas antes', hoteles.shape[0])
print('Filas despues', hoteles_limpio.shape[0])
print('Filas eliminadas', hoteles.shape[0] - hoteles_limpio.shape[0])

Filas antes 119390
Filas despues 79225
Filas eliminadas 40165


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [ ]:
# Escribe tu código aquí
# Sacamos el promedio de ambas columnas agrupadas con el df original
resumen_original = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean()
resumen_original = resumen_original.reset_index()
# Sacamos el promedio de ambas columnas con el df sin duplicados
resumen_limpio = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean()
resumen_limpio = resumen_limpio.reset_index()
print(resumen_original)
print(resumen_limpio)


          hotel  is_canceled         adr
0    City Hotel     0.417270  105.304465
1  Resort Hotel     0.277634   94.952930
          hotel  is_canceled         adr
0    City Hotel     0.286558  112.564402
1  Resort Hotel     0.225579  100.495582


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [ ]:
# Escribe tu código aquí
# le agreagamos las columnas a cada df
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'
# Concatenamos ambos resumenes
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)
# Acomodamos por hotel
comparacion = comparacion.sort_values(by=['hotel'])
comparacion = comparacion.reset_index(drop=True)
print(comparacion)

          hotel  is_canceled         adr         version
0    City Hotel     0.417270  105.304465        original
1    City Hotel     0.286558  112.564402  sin duplicados
2  Resort Hotel     0.277634   94.952930        original
3  Resort Hotel     0.225579  100.495582  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

1.- La metrica que mas cambio fue la de is canceled, se nota como disminuye drasticamente cuando se eliminan los duplicados

2.- Podrian inclinar la balanza hacia falsos positivos por tanta fila duplicada

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [ ]:
# Escribe tu código aquí
# Creamos un nuevo dataframe con el limpio
dubs = hoteles_limpio.copy()
dubs['es_duplicado'] = dubs.duplicated()
# Verificamos cuantos duplicados suman
print(dubs['es_duplicado'].value_counts())
print(dubs['es_duplicado'].sum())

es_duplicado
False    79225
Name: count, dtype: int64
0


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [ ]:
# Escribe tu código aquí
df_h3 = hoteles_limpio.copy()
col = df_h3.columns.drop("reservation_status_date")
# Eliminamos los duplicados de las demas columnas sin contar el reservation date status
df_h3 = df_h3.drop_duplicates(subset=col, keep='last').copy()
# Sacamos cuantas filas fueron eliminadas
filas = hoteles_limpio.shape[0] - df_h3.shape[0]
# Acomodamos el df
df_h3 = df_h3.sort_values(by=['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date'])
# Imprimimos
print('Filas sobrantes:',filas)
print('Filas ordenadas:\n',df_h3.head(10))
print('Distribucion de Reservation status:\n',df_h3['reservation_status'].value_counts())

Filas sobrantes: 106
Filas ordenadas:
             hotel  is_canceled  lead_time  arrival_date_year  \
31842  City Hotel            0          0               2015   
31855  City Hotel            0          0               2015   
31857  City Hotel            0          0               2015   
47961  City Hotel            0          0               2015   
31870  City Hotel            0          0               2015   
48053  City Hotel            0          0               2015   
31869  City Hotel            0          0               2015   
31871  City Hotel            0          0               2015   
31872  City Hotel            0          0               2015   
31881  City Hotel            0          0               2015   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
31842             August                        31                          1   
31855             August                        32                          3   
31857        

**Respuesta 4.2:**

Por lo que podemos ver, aun hubieron 106 filas mas que pudieran parecer duplicadas, solo que la diferencia la marcada la fecha del estatus de la reservacion, estas pudieras ser registros de un programa que cada cierto tiempo corre algun cron job que se asegura que nada haya cambiado, y solo actualiza la ultima del status confirmando que sigue igual, creo que se pudieran eliminar ya que no representarian un cambio real en los datos, solo se estaria duplicando informacion que pudiera meter ruido mas adelante

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Supongo que es hablar solamente de lo que hicimos en el punto 4...
Primero verificamos que realmente no hubieran filas duplicadas, y en el siguiente punto eliminamos las columnas duplicadas, sin tomar en cuenta la columna de fecha de estatus de reservacion, con la regla de conservacion keep=last el cual conservaba el ultimo registro de los duplicados como lo pide el punto 4.2, por lo cual se evaluaron las columnas: hotel, lead_time, arrival_date_year, arrival_date_month, arrival_date_day_of_month, adr, reservation_status, reservation_status_date.
Obteniendo que se eliminaron 106 filas duplicadas

Supuestos y limitaciones, considerando que nunca organizamos el DF por fechas, no podemos asegurar que se hayan mantenido solamente los registros mas recientes de los duplicados, al no tener reservation id o algo parecido, nos limita un poco el como podemos asegurarnos de que sean datos reales o datos repetidos, ya que no poseen un dato de identidad unico

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1.- Para medir la eficiencia del sistema o del capturista, los duplicados nos pueden ayudar a entender que posiblemente haya una oportunidad de mejora en el sistema, si os eliminamos, creeriamos que el sistema trabaja perfecto, cuando no es asi

2.- Un id de cliente, asi habria otro comparativo para saber si la fila esta duplicada o no, ya que pudiera existir los registros donde dos personas se iban a quedar en la misma fecha, y todos los demas datos de la columna ser iguales, y la unica diferencia notable seria el id del cliente, el saber que no fue una doble captura, sino dos clientes muy parecidos el mismo dia.

3.- Puede ser por varias razones, si no hemos normalizados nuestros datos, duplicated no veria diferencia si en un registro existe Check-Out y en otro esta escrito como check-Out, lo cual si seria un duplicado, o ya sea que como lo vimos en el ultimo ejercicio pueden ser identicos en 9 de 10 columnas y la ultima no ser de relevancia, pero si no son duplicados exactos duplicated no lo tomara en cuenta.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.